unexpected panic while compiling firefox on arm64

Hi, I see this strange error while trying to compile firefox-59.0.1 with rust-1.23 on arm64 laptop running gentoo linux. It says, my bug report would be appreciated. I'd also appreciate any feedback.

I would be happy to provide any additional information.

```
   Compiling quote v0.3.15
   Compiling peeking_take_while v0.1.2
   Compiling glob v0.2.11
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.23.0-dev running on aarch64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'other was less than the current instant', src/libstd/sys/unix/time.rs:284:16
stack backtrace:
   0:       0x7f9e7de973 - rust_metadata_std_5958eddae63d6947e0733521fc5435ed
   1:       0x7f9e7da9d7 - rust_metadata_std_5958eddae63d6947e0733521fc5435ed
   2:       0x7f9e7e9f3f - rust_metadata_std_5958eddae63d6947e0733521fc5435ed
   3:       0x7f9e7e9c4f - rust_metadata_std_5958eddae63d6947e0733521fc5435ed
   4:       0x7f9e7ea443 - std::panicking::rust_panic_with_hook::h22b280fc4fc0de87
   5:       0x7f9e7ea303 - rust_metadata_std_5958eddae63d6947e0733521fc5435ed
   6:       0x7f9e7d9eb7 - std::time::Instant::elapsed::hf802fe319eded9b8
   7:       0x7f9da24283 - <unknown>
   8:       0x7f9da4965b - <unknown>
   9:       0x7f9cd3199b - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  10:       0x7f9cf028f7 - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  11:       0x7f9cc1cd2b - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  12:       0x7f9d09b153 - <unknown>
  13:       0x7f9cd31a33 - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  14:       0x7f9cd320ff - rustc::ty::maps::<impl rustc::ty::maps::queries::symbol_name<'tcx>>::try_get::hf71dec19792cb774
  15:       0x7f9ce83073 - rustc::ty::maps::TyCtxtAt::symbol_name::hf81eb960dfd03136
  16:       0x7f9d0ae4f3 - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::symbol_name::hd40df66149759931
  17:       0x7f9d984003 - <unknown>
  18:       0x7f9d9609c7 - <unknown>
  19:       0x7f9d984fc3 - <unknown>
  20:       0x7f9cd7a433 - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  21:       0x7f9cf0ad07 - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  22:       0x7f9cc1e81b - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  23:       0x7f9d081dbb - <unknown>
  24:       0x7f9cd7a4af - rust_metadata_rustc_3ea30897d9dddf7b5fa3a6c2523637e7
  25:       0x7f9cd7a9df - rustc::ty::maps::<impl rustc::ty::maps::queries::exported_symbols<'tcx>>::try_get::h7cd7159316d987b4
  26:       0x7f9ce870e3 - rustc::ty::maps::TyCtxtAt::exported_symbols::h35b6eb150f9aa7b2
  27:       0x7f9d0af0df - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::exported_symbols::h69dd1a8087b4ddb7
  28:       0x7f9da0be0f - <unknown>
  29:       0x7f9da0b253 - rustc_trans::back::write::start_async_translation::hd810efc5779dac71
  30:       0x7f9da4ea93 - rustc_trans::base::trans_crate::h1ab479cd2e56d359
  31:       0x7f9d9f930f - <rustc_trans::LlvmTransCrate as rustc_trans_utils::trans_crate::TransCrate>::trans_crate::hf0c11cb959ccfc0e
  32:       0x7f9e9b987f - <unknown>
  33:       0x7f9e949f1b - <unknown>
  34:       0x7f9e9a949f - <unknown>
  35:       0x7f9e9a1967 - <unknown>
  36:       0x7f9e99ed5f - <unknown>
  37:       0x7f9e910633 - <unknown>
  38:       0x7f9e9462ff - rustc_driver::driver::compile_input::h10487c1bd068aca0
  39:       0x7f9e9952bb - rustc_driver::run_compiler::h90c48c1cfea93e36
  40:       0x7f9e92682b - <unknown>
  41:       0x7f9e7f5d17 - __rust_maybe_catch_panic
  42:       0x7f9e988733 - <unknown>
  43:       0x7f9e7e8e6f - rust_metadata_std_5958eddae63d6947e0733521fc5435ed
  44:       0x7f9c3f25bb - <unknown>

error: Could not compile `glob`.
warning: build failed, waiting for other jobs to finish...
error: build failed
gmake[4]: *** [/var/tmp/portage/www-client/firefox-59.0.1/work/mozilla-release-3db9e3d52b17563efca181ccbb50deb8660c59ae/config/rules.mk:972: force-cargo-library-build] Error 101
gmake[3]: *** [/var/tmp/portage/www-client/firefox-59.0.1/work/mozilla-release-3db9e3d52b17563efca181ccbb50deb8660c59ae/config/recurse.mk:73: toolkit/library/rust/target] Error 2
gmake[2]: *** [/var/tmp/portage/www-client/firefox-59.0.1/work/mozilla-release-3db9e3d52b17563efca181ccbb50deb8660c59ae/config/recurse.mk:33: compile] Error 2
gmake[1]: *** [/var/tmp/portage/www-client/firefox-59.0.1/work/mozilla-release-3db9e3d52b17563efca181ccbb50deb8660c59ae/config/rules.mk:434: default] Error 2
gmake: *** [client.mk:168: build] Error 2
0 compiler warnings present.
```

--- a/src/librustc/util/profiling.rs
+++ b/src/librustc/util/profiling.rs
@@ -2,7 +2,7 @@ use session::config::Options;
 
 use std::fs;
 use std::io::{self, StderrLock, Write};
-use std::time::{Duration, Instant};
+use std::time::Instant;
 
 macro_rules! define_categories {
     ($($name:ident,)*) => {
@@ -203,20 +203,7 @@ impl SelfProfiler {
     }
 
     fn stop_timer(&mut self) -> u64 {
-        let elapsed = if cfg!(windows) {
-            // On Windows, timers don't always appear to be monotonic (see #51648)
-            // which can lead to panics when calculating elapsed time.
-            // Work around this by testing to see if the current time is less than
-            // our recorded time, and if it is, just returning 0.
-            let now = Instant::now();
-            if self.current_timer >= now {
-                Duration::new(0, 0)
-            } else {
-                self.current_timer.elapsed()
-            }
-        } else {
-            self.current_timer.elapsed()
-        };
+        let elapsed = self.current_timer.elapsed();
 
         self.current_timer = Instant::now();
 

--- a/src/libstd/sys/cloudabi/time.rs
+++ b/src/libstd/sys/cloudabi/time.rs
@@ -25,6 +25,14 @@ impl Instant {
         }
     }
 
+    pub fn actually_monotonic() -> bool {
+        true
+    }
+
+    pub const fn zero() -> Instant {
+        Instant { t: 0 }
+    }
+
     pub fn sub_instant(&self, other: &Instant) -> Duration {
         let diff = self.t
             .checked_sub(other.t)

--- a/src/libstd/sys/redox/time.rs
+++ b/src/libstd/sys/redox/time.rs
@@ -128,6 +128,14 @@ impl Instant {
         Instant { t: now(syscall::CLOCK_MONOTONIC) }
     }
 
+    pub const fn zero() -> Instant {
+        Instant { t: Timespec { t: syscall::TimeSpec { tv_sec: 0, tv_nsec: 0 } } }
+    }
+
+    pub fn actually_monotonic() -> bool {
+        false
+    }
+
     pub fn sub_instant(&self, other: &Instant) -> Duration {
         self.t.sub_timespec(&other.t).unwrap_or_else(|_| {
             panic!("specified instant was later than self")

--- a/src/libstd/sys/unix/time.rs
+++ b/src/libstd/sys/unix/time.rs
@@ -14,6 +14,12 @@ struct Timespec {
 }
 
 impl Timespec {
+    const fn zero() -> Timespec {
+        Timespec {
+            t: libc::timespec { tv_sec: 0, tv_nsec: 0 },
+        }
+    }
+
     fn sub_timespec(&self, other: &Timespec) -> Result<Duration, Duration> {
         if self >= other {
             Ok(if self.t.tv_nsec >= other.t.tv_nsec {
@@ -128,17 +134,20 @@ mod inner {
     }
 
     pub const UNIX_EPOCH: SystemTime = SystemTime {
-        t: Timespec {
-            t: libc::timespec {
-                tv_sec: 0,
-                tv_nsec: 0,
-            },
-        },
+        t: Timespec::zero(),
     };
 
     impl Instant {
         pub fn now() -> Instant {
             Instant { t: unsafe { libc::mach_absolute_time() } }
+        }
+
+        pub const fn zero() -> Instant {
+            Instant { t: 0 }
+        }
+
+        pub fn actually_monotonic() -> bool {
+            true
         }
 
         pub fn sub_instant(&self, other: &Instant) -> Duration {
@@ -258,17 +267,24 @@ mod inner {
     }
 
     pub const UNIX_EPOCH: SystemTime = SystemTime {
-        t: Timespec {
-            t: libc::timespec {
-                tv_sec: 0,
-                tv_nsec: 0,
-            },
-        },
+        t: Timespec::zero(),
     };
 
     impl Instant {
         pub fn now() -> Instant {
             Instant { t: now(libc::CLOCK_MONOTONIC) }
+        }
+
+        pub const fn zero() -> Instant {
+            Instant {
+                t: Timespec::zero(),
+            }
+        }
+
+        pub fn actually_monotonic() -> bool {
+            (cfg!(target_os = "linux") && cfg!(target_arch = "x86_64")) ||
+            (cfg!(target_os = "linux") && cfg!(target_arch = "x86")) ||
+            false // last clause, used so `||` is always trailing above
         }
 
         pub fn sub_instant(&self, other: &Instant) -> Duration {

--- a/src/libstd/sys/wasm/time.rs
+++ b/src/libstd/sys/wasm/time.rs
@@ -14,6 +14,14 @@ impl Instant {
         Instant(TimeSysCall::perform(TimeClock::Monotonic))
     }
 
+    pub const fn zero() -> Instant {
+        Instant(Duration::from_secs(0))
+    }
+
+    pub fn actually_monotonic() -> bool {
+        false
+    }
+
     pub fn sub_instant(&self, other: &Instant) -> Duration {
         self.0 - other.0
     }

--- a/src/libstd/sys/windows/time.rs
+++ b/src/libstd/sys/windows/time.rs
@@ -40,6 +40,14 @@ impl Instant {
         t
     }
 
+    pub fn actually_monotonic() -> bool {
+        false
+    }
+
+    pub const fn zero() -> Instant {
+        Instant { t: 0 }
+    }
+
     pub fn sub_instant(&self, other: &Instant) -> Duration {
         // Values which are +- 1 need to be considered as basically the same
         // units in time due to various measurement oddities, according to

--- a/src/libstd/time.rs
+++ b/src/libstd/time.rs
@@ -12,11 +12,13 @@
 
 #![stable(feature = "time", since = "1.3.0")]
 
+use cmp;
 use error::Error;
 use fmt;
 use ops::{Add, Sub, AddAssign, SubAssign};
 use sys::time;
 use sys_common::FromInner;
+use sys_common::mutex::Mutex;
 
 #[stable(feature = "time", since = "1.3.0")]
 pub use core::time::Duration;
@@ -150,7 +152,45 @@ impl Instant {
     /// ```
     #[stable(feature = "time2", since = "1.8.0")]
     pub fn now() -> Instant {
-        Instant(time::Instant::now())
+        let os_now = time::Instant::now();
+
+        // And here we come upon a sad state of affairs. The whole point of
+        // `Instant` is that it's monotonically increasing. We've found in the
+        // wild, however, that it's not actually monotonically increasing for
+        // one reason or another. These appear to be OS and hardware level bugs,
+        // and there's not really a whole lot we can do about them. Here's a
+        // taste of what we've found:
+        //
+        // * #48514 - OpenBSD, x86_64
+        // * #49281 - linux arm64 and s390x
+        // * #51648 - windows, x86
+        // * #56560 - windows, x86_64, AWS
+        // * #56612 - windows, x86, vm (?)
+        // * #56940 - linux, arm64
+        // * https://bugzilla.mozilla.org/show_bug.cgi?id=1487778 - a similar
+        //   Firefox bug
+        //
+        // It simply seems that this it just happens so that a lot in the wild
+        // we're seeing panics across various platforms where consecutive calls
+        // to `Instant::now`, such as via the `elapsed` function, are panicking
+        // as they're going backwards. Placed here is a last-ditch effort to try
+        // to fix things up. We keep a global "latest now" instance which is
+        // returned instead of what the OS says if the OS goes backwards.
+        //
+        // To hopefully mitigate the impact of this though a few platforms are
+        // whitelisted as "these at least haven't gone backwards yet".
+        if time::Instant::actually_monotonic() {
+            return Instant(os_now)
+        }
+
+        static LOCK: Mutex = Mutex::new();
+        static mut LAST_NOW: time::Instant = time::Instant::zero();
+        unsafe {
+            let _lock = LOCK.lock();
+            let now = cmp::max(LAST_NOW, os_now);
+            LAST_NOW = now;
+            Instant(now)
+        }
     }
 
     /// Returns the amount of time elapsed from another instant to this one.

